sdk/appcenter-analytics/src/main/java/com/microsoft/appcenter/analytics/EventProperties.java

```diff
@@ -0,0 +1,156 @@
+package com.microsoft.appcenter.analytics;
+
+import com.microsoft.appcenter.ingestion.models.properties.BooleanTypedProperty;
+import com.microsoft.appcenter.ingestion.models.properties.DateTimeTypedProperty;
+import com.microsoft.appcenter.ingestion.models.properties.DoubleTypedProperty;
+import com.microsoft.appcenter.ingestion.models.properties.LongTypedProperty;
+import com.microsoft.appcenter.ingestion.models.properties.StringTypedProperty;
+import com.microsoft.appcenter.ingestion.models.properties.TypedProperty;
+import com.microsoft.appcenter.utils.AppCenterLog;
+
+import java.util.Date;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static com.microsoft.appcenter.analytics.Analytics.LOG_TAG;
+
+/**
+ * Event properties builder.
+ */
+public class EventProperties {
+
+    private static final String VALUE_NULL_ERROR_MESSAGE = "Property value cannot be null";
+
+    /**
+     * Properties key/value pairs.
+     */
+    private final Map<String, TypedProperty> mProperties = new LinkedHashMap<>();
+
+    Map<String, TypedProperty> getProperties() {
```

Comment:
Why make this a `Map` when it's converted to an `ArrayList` when it's used elsewhere in the SDK?

Reply:
We use Map in the EventProperties to track duplicates and print an error.
Once the user input has been submitted, we need to make a copy to avoid concurrent modification exceptions and don't need to track duplicate anymore, since the schema requires a list, we take the need to copy opportunity to keep as a list to minimize copies and allocations.